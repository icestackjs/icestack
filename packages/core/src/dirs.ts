import path from 'node:path'

export function createResolveDir(base: string) {
  const rootDir = path.dirname(require.resolve(`${base}/package.json`))
  const assetsDir = path.resolve(rootDir, 'assets')
  const scssDir = path.resolve(assetsDir, 'scss')
  const jsDir = path.resolve(assetsDir, 'js')
  const cvaDir = path.resolve(assetsDir, 'cva')
  const cssDir = path.resolve(assetsDir, 'css')
  const cssResolvedDir = path.resolve(assetsDir, 'css-resolved')
  const pluginsDir = path.resolve(assetsDir, 'plugins')

  function resolveCssDir(dir?: string) {
    return dir ? path.resolve(dir, 'css') : cssDir
  }
  function resolveScssDir(dir?: string) {
    return dir ? path.resolve(dir, 'scss') : scssDir
  }

  function resolveCssResolvedDir(dir?: string) {
    return dir ? path.resolve(dir, 'css-resolved') : cssResolvedDir
  }

  function resolveJsDir(dir?: string) {
    return dir ? path.resolve(dir, 'js') : jsDir
  }

  function resolveCvaDir(dir?: string) {
    return dir ? path.resolve(dir, 'cva') : cvaDir
  }

  function getCssPath(relPath: string, dir?: string) {
    const cssDir = resolveCssDir(dir)
    const cssPath = path.resolve(cssDir, relPath)
    return cssPath + '.css'
  }

  function getScssPath(relPath: string, dir?: string) {
    const scssDir = resolveScssDir(dir)
    const cssPath = path.resolve(scssDir, relPath)
    return cssPath + '.scss'
  }

  function getCssResolvedPath(relPath: string, dir?: string) {
    const cssDir = resolveCssResolvedDir(dir)
    const cssPath = path.resolve(cssDir, relPath)
    return cssPath + '.css'
  }

  function getJsPath(relPath: string, dir?: string) {
    const targetJsDir = resolveJsDir(dir)
    const jsPath = path.resolve(targetJsDir, relPath)
    return jsPath + '.cjs'
  }

  function getCvaPath(relPath: string, dir?: string, format: string = 'ts') {
    const targetJsDir = resolveJsDir(dir)
    const jsPath = path.resolve(targetJsDir, relPath)
    return jsPath + '.' + format
  }

  function getPluginsPath(relPath: string, dir?: string) {
    const targetJsDir = dir ? path.resolve(dir, 'js') : pluginsDir
    const jsPath = path.resolve(targetJsDir, relPath)
    return jsPath + '.cjs'
  }
  return {
    assetsDir,
    scssDir,
    jsDir,
    cssDir,
    cssResolvedDir,
    pluginsDir,
    getCssPath,
    getJsPath,
    getPluginsPath,
    getCssResolvedPath,
    resolveCssDir,
    resolveCssResolvedDir,
    resolveJsDir,
    getScssPath,
    resolveCvaDir,
    getCvaPath
  }
}
