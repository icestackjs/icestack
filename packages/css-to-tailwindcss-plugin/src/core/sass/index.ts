import type { Options, StringOptions } from 'sass'

export function tryLoadSass() {
  return import('sass')
}
export function tryLoadSassSync() {
  return require('sass')
}

export function isExtSassFile(extname: string) {
  return extname === '.scss' || extname === '.sass'
}

export async function sassCompile(path: string, sassOptions?: Options<'sync'>) {
  const { compile } = await tryLoadSass()
  const { css } = compile(path, sassOptions)
  return css
}

export function sassCompileSync(path: string, sassOptions?: Options<'sync'>) {
  const { compile } = tryLoadSassSync()
  const { css } = compile(path, sassOptions)
  return css
}

export async function sassCompileString(code: string, sassOptions?: StringOptions<'sync'>) {
  const { compileString } = await tryLoadSass()
  const { css } = compileString(code, sassOptions)
  return css
}
