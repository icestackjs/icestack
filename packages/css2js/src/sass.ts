import { RawCssInput } from './types'

export function tryLoadSass() {
  return import('sass')
}

export function isExtSassFile(extname: string) {
  return extname === '.scss' || extname === '.sass'
}

export function isLangSass(lang: RawCssInput['lang']) {
  return lang === 'scss' || lang === 'sass'
}
