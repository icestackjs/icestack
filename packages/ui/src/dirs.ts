import path from 'node:path'
import { pkgName } from './constants'
//  require.resolve('@icestack/ui/package.json')
const assetsDir = path.resolve(path.dirname(require.resolve(`${pkgName}/package.json`)), 'assets')
const scssDir = path.resolve(assetsDir, 'scss')
const jsDir = path.resolve(assetsDir, 'js')
const cssDir = path.resolve(assetsDir, 'css')
const cssResolvedDir = path.resolve(assetsDir, 'css-resolved')
const pluginsDir = path.resolve(assetsDir, 'plugins')

const componentTemplate = path.resolve(scssDir, 'components/t.scss')

function resolveCssDir(dir?: string) {
  return dir ? path.resolve(dir, 'css') : cssDir
}

function resolveCssResolvedDir(dir?: string) {
  return dir ? path.resolve(dir, 'css-resolved') : cssResolvedDir
}

function resolveJsDir(dir?: string) {
  return dir ? path.resolve(dir, 'js') : jsDir
}

function getCssPath(relPath: string, dir?: string) {
  const targetCssDir = resolveCssDir(dir)
  const cssPath = path.resolve(targetCssDir, relPath)
  return cssPath.replace(/scss$/, 'css')
}

function getCssResolvedpath(relPath: string, dir?: string) {
  const targetCssDir = resolveCssResolvedDir(dir)
  const jsPath = path.resolve(targetCssDir, relPath)
  return jsPath.replace(/scss$/, 'css')
}

function getJsPath(relPath: string, dir?: string) {
  const targetJsDir = resolveJsDir(dir)
  const jsPath = path.resolve(targetJsDir, relPath)
  return jsPath.replace(/scss$/, 'js')
}

function getPluginsPath(relPath: string, dir?: string) {
  const targetJsDir = dir ? path.resolve(dir, 'js') : pluginsDir
  const jsPath = path.resolve(targetJsDir, relPath)
  return jsPath.replace(/scss$/, 'js')
}

export {
  assetsDir,
  scssDir,
  jsDir,
  cssDir,
  cssResolvedDir,
  pluginsDir,
  getCssPath,
  getJsPath,
  getPluginsPath,
  getCssResolvedpath,
  resolveCssDir,
  resolveCssResolvedDir,
  resolveJsDir,
  componentTemplate
}
