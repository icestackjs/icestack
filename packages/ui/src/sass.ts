import { merge, parse } from '@icestack/postcss-utils'
export * from '@icestack/scss'

export function mergeRoot(arr: string[]) {
  return merge(...arr.map((x) => parse(x)))
}
