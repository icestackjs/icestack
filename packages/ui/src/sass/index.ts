export * from './functions'
export * from './utils'
