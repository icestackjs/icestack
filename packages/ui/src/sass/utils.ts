import { OrderedMap } from 'immutable'
import { TinyColor } from '@ctrl/tinycolor'
import * as sass from 'sass'
import defu from 'defu'

export interface TransformJsToSassOptions {
  quotes?: boolean
  denominatorUnits?: string[]
  numeratorUnits?: string[]
  radix?: number
}
// : sass.Value | sass.Value[] | sass.SassList | sass.SassMap
export function transformJsToSass(raw: any, options?: TransformJsToSassOptions): sass.Value {
  const { denominatorUnits, numeratorUnits, quotes, radix } = defu<Required<TransformJsToSassOptions>, TransformJsToSassOptions[]>(options, {
    quotes: false
  })
  if (raw === undefined || raw === null) {
    return sass.sassNull
  } else if (typeof raw === 'string') {
    // default quotes false
    return new sass.SassString(raw, {
      quotes
    })
  } else if (typeof raw === 'boolean') {
    return raw ? sass.sassTrue : sass.sassFalse
  } else if (typeof raw === 'number') {
    return new sass.SassNumber(raw, {
      denominatorUnits,
      numeratorUnits
    })
  } else if (typeof raw === 'bigint') {
    return new sass.SassString(raw.toString(radix), {
      quotes
    })
  } else if (Array.isArray(raw)) {
    return new sass.SassList(
      raw.map((x) => {
        return transformJsToSass(x)
      })
    )
  } else if (typeof raw === 'object') {
    return new sass.SassMap(
      OrderedMap(
        Object.entries(raw).map(([key, value]) => {
          return [transformJsToSass(key), transformJsToSass(value)]
        })
      )
    )
  } else if (typeof raw === 'function') {
    return typeof raw.signature === 'string' ? new sass.SassFunction(raw.signature, raw) : sass.sassNull
  } else {
    return sass.sassNull
  }
}

export function transformBaseJs(entries: [string, string][]): [sass.Value, sass.Value][] {
  return entries.map(([varName, value]) => {
    const color = new TinyColor(value)
    let v: sass.Value
    if (color.isValid) {
      let str = ''
      str = color.a < 1 && color.a > 0 ? `${color.r} ${color.g} ${color.b} / ${color.a}` : `${color.r} ${color.g} ${color.b}`
      v = new sass.SassString(str, {
        quotes: false
      })
    } else {
      v = new sass.SassString(value, {
        quotes: false
      })
    }
    return [
      new sass.SassString(varName, {
        quotes: false
      }),
      v
    ]
  })
}
