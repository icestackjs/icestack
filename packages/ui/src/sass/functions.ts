import type { Options } from 'sass'
import { Value, SassMap } from 'sass'
import { get } from 'lodash'
import { OrderedMap } from 'immutable'
import type allComponents from '../allComponents'
import { transformBaseJs, transformJsToSass } from './utils'
import * as avatar from '@/components/avatar'
import * as bottomNavigation from '@/components/bottom-navigation'
import * as button from '@/components/button'
import * as alert from '@/components/alert'
import * as badge from '@/components/badge'
import * as chat from '@/components/chat'
import * as checkbox from '@/components/checkbox'
import * as fileInput from '@/components/file-input'
import * as input from '@/components/input'
import * as link from '@/components/link'
import * as progress from '@/components/progress'
import * as textarea from '@/components/textarea'
import { defaultVarsEntries } from '@/base/css-vars'

// @ts-ignore
const defaultPreset: Record<(typeof allComponents)[number], any> = {
  alert: alert.options,
  avatar: avatar.options,
  bottomNavigation: bottomNavigation.options,
  button: button.options,
  badge: badge.options,
  chat: chat.options,
  checkbox: checkbox.options,
  fileInput: fileInput.options,
  input: input.options,
  link: link.options,
  progress: progress.options,
  textarea: textarea.options
}

export const functions: Options<'sync'>['functions'] = {
  'injectCssVars()': () => {
    return new SassMap(OrderedMap(transformBaseJs(defaultVarsEntries)))
  },
  'globalAtMediaHover()': () => {
    return transformJsToSass(false)
  },
  'inject($path:null)': (args: Value[]) => {
    const p = args[0].assertString().text
    const map = get(defaultPreset, p, {})
    return transformJsToSass(map)
  }
}
