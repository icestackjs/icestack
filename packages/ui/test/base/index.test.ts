import path from 'node:path'
import { compileScss } from '@/sass'
import { scssDir } from '@/dirs'

export function resolve(filename: string) {
  return path.resolve(scssDir, 'base', filename + '.scss')
}

describe('base', () => {
  it('snap', async () => {
    const css = await compileScss(resolve('index'))
    expect(css).toMatchSnapshot()
  })
})
