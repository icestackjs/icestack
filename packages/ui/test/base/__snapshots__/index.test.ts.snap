// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`base > snap 1`] = `
":root {
  --ice-primary: 22 119 255;
  --ice-primary-focus: 64 150 255;
  --ice-primary-content: 255 255 255;
  --ice-success-content: 255 255 255;
  --ice-success: 82 196 26;
  --ice-success-focus: 115 209 61;
  --ice-warning-content: 255 255 255;
  --ice-warning: 250 173 20;
  --ice-warning-focus: 255 197 61;
  --ice-error-content: 255 255 255;
  --ice-error: 245 34 45;
  --ice-error-focus: 255 77 79;
  --ice-info: 22 119 255;
  --ice-info-content: 255 255 255;
  --ice-info-focus: 64 150 255;
  --ice-neutral: 191 191 191;
  --ice-neutral-content: 255 255 255;
  --ice-neutral-focus: 217 217 217;
  --ice-base-100: 255 255 255;
  --ice-base-200: 242 242 242;
  --ice-base-300: 229 230 230;
  --ice-base-content: 31 41 55;
  --ice-rounded-box: 1rem;
  --ice-rounded-btn: 0.5rem;
  --ice-rounded-badge: 1.9rem;
  --ice-animation-btn: 0.25s;
  --ice-animation-input: 0.2s;
  --ice-btn-text-case: uppercase;
  --ice-btn-focus-scale: 0.95;
  --ice-border-btn: 1px;
  --ice-tab-border: 1px;
  --ice-tab-radius: 0.5rem;
}"
`;
